### PR TITLE
executor: add recover mechanism for index join workers

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

**Comment by zz-jason** [comment]
@tiancaiamao @coocood PTAL

File: executor/index_lookup_join.go [hunk 1]
```diff
@@ -355,21 +376,30 @@ func (ow *outerWorker) increaseBatchSize() {
...
-			task.doneCh <- errors.Trace(err)
-			if err != nil {
-				return
-			}
 		case <-ctx.Done():
 			return
 		}
+
+		err := iw.handleTask(ctx, task)
+		task.doneCh <- errors.Trace(err)
```
Comment (tiancaiamao) [inline executor/index_lookup_join.go]: I wonder is it really necessary to send `nil` to `taskk.doneCh` ?

Comment (zz-jason) [reply executor/index_lookup_join.go]: We can close `task.doneCh` when `err == nil` actually, but send a `nil` is simpler.

File: executor/index_lookup_join.go [hunk 2]
```diff
@@ -291,24 +296,40 @@ func (ow *outerWorker) run(ctx context.Context, wg *sync.WaitGroup) {
...
+			ow.pushToResultChan(ctx, task)
 			return
 		}
 		if task == nil {
 			return
 		}
-		select {
-		case <-ctx.Done():
+
+		if finished := ow.pushToInnerChan(ctx, task); finished {
```
[the remark this answers is not in the document]
Comment (zz-jason) [reply executor/index_lookup_join.go]: The order does not affect correctness. I think is better to push `task` to `innerCh` to make inner worker works as soon as possible.

File: executor/index_lookup_join.go [hunk 3]
```diff
@@ -291,24 +296,40 @@ func (ow *outerWorker) run(ctx context.Context, wg *sync.WaitGroup) {
...
-		select {
-		case <-ctx.Done():
+		if finished := ow.pushToResultChan(ctx, task); finished {
 			return
-		case ow.innerCh <- task:
 		}
 	}
 }
 
+func (ow *outerWorker) pushToResultChan(ctx context.Context, task *lookUpJoinTask) bool {
```
Comment (coocood) [inline executor/index_lookup_join.go]: How about making the channel an argument and remove the other push function?

File: executor/index_lookup_join.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply executor/index_lookup_join.go]: Get it.

**Comment by tiancaiamao** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**